<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">

<html>
  <head>
    <title>References</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
    <meta name="description" content=""/>
    <meta name="generator" content="ParaDocs - www.paradocsapp.com" />
    <meta name="keywords" content="references"/>
    <link href="../help.css" media="screen" rel="stylesheet" type="text/css" />
  </head>
  <body>
  	<div='page'>
		<div id='headerIconAndText'>
			<table><tr><td>
				<a href='../index.html'><img width='80' height='80' src='../images/RevBayesIcon' style='float:left; margin: 0 5px 0 0; vertical-align:middle;' /></a>
			</td><td>
				<h1 style='vertical-align:middle'>References</h1>
			</td></tr></table>
		</div>
  	
  		<div id='bodyText'>
    		<ul>
<li><p>Do, CB, MSP Mahabhashyam, M Brudno, and S Batzoglou. 2005. PROBCONS: Probabilistic consistency-based multiple sequence alignment. <em>Genome Research</em> 15:330-340.</p></li>
<li><p>Drummond, AJ, MA Suchard, D Xie, and A Rambaut. 2012. Bayesian phylogenetics with BEAUti and the BEAST 1.7 <em>Molecular Biology And Evolution</em> 29:1969-1973.</p></li>
<li><p>Edgar, RC. 2004. MUSCLE: Multiple sequence alignment with high accuracy and high throughput <em>Nucleic Acids Research</em> 32:1792-1797.</p></li>
<li><p>Efron, B. 1979. Bootstrap methods: Another look at the jackknife. <em>The Annals of Statistics</em> 7:1-26.</p></li>
<li><p>Felsenstein, J. 1985. Confidence limits on phylogenies: An approach using the bootstrap. <em>Evolution</em> 39:783-791.</p></li>
<li><p>Huelsenbeck, JP, and F Ronquist. 2001. MrBayes: Bayesian inference of phylogeny. <em>Bioinformatics</em> 17:754-755.</p></li>
<li><p>Ronquist, F, and JP Huelsenbeck. 2003. MrBayes version 3.0: Bayesian phylogenetic inference under mixed models. <em>Bioinformatics</em> 19:1572-1574.</p></li>
<li><p>Höhna, S, TA Heath, B Boussau, MJ Landis, F Ronquist, and JP Huelsenbeck. 2014. Probabilistic graphical model representation in phylogenetics. <em>Systematic Biology</em> 63:753-771.</p></li>
<li><p>Höhna, S, MJ Landis, TA Heath, B Boussau, N Lartillot, BR Moore, JP Huelsenbeck, and F Ronquist. 2016. RevBayes: Bayesian phylogenetic inference using graphical models and an interactive model-specification language. <em>Systematic Biology</em> 65:726-736.</p></li>
<li><p>Larkin MA, G Blackshields, NP Brown, R Chenna, PA McGettigan, H McWilliam, F Valentin, IM Wallace, A Wilm, R Lopez, JD Thompson, TJ Gibson, and DG Higgins. 2007. <em>Bioinformatics</em> 23:2947-2948.</p></li>
<li><p>Lewis, PO. 2003. NCL: A C++ class library for interpreting data files in NEXUS format. <em>Bioinformatics</em> 19:2330-2331.</p></li>
<li><p>Maddison, WP and DR Maddison. 2016. Mesquite: a modular system for evolutionary analysis.  Version 3.10  <a href="http://mesquiteproject.org" title="Mesquite Website">Mesquite Project</a></p></li>
<li><p>Maddison, DR, DL Swofford, and WP Maddison. 1997. NEXUS: An extensible file format for systematic information. <em>Systematic Biology</em> 46:590-621.</p></li>
<li><p>Notredame, C, DG Higgins, and J Heringa. 2000. T-Coffee: A novel method for fast and accurate multiple sequence alignment. <em>Journal of Molecular Biology</em> 302:205-217.</p></li>
<li><p>Swofford, DL. 2003. <em>PAUP*. Phylogenetic Analysis Using Parsimony (*and Other Methods)</em>. Version 4. Sinauer Associates, Sunderland, Massachusetts.</p></li>
<li><p>Swofford, DL, GJ Olsen, PJ Waddell, and DM Hillis. 1996. Phylogenetic Inference. In DM Hillis, C Moritz, and BK Mable (Eds.) <em>Molecular Systematics</em>. Sinauer ssociates, Sunderland.</p></li>
</ul>

    	</div>
    	
    	
    	
    </div>
  </body>
</html>
